File: Web Services and Middleware Technologies/Web Services/CSharpServer/CSharpServer/repository/SubjectRepository.cs
using System.Collections.Generic;
using System.Linq;
using CSharpServer.config;
using CSharpServer.domain;

namespace CSharpServer.repository
{
    public class SubjectRepository
    {
        public SubjectRepository()
        {
        }

        public void Add(Subject subject)
        {
            using var context = new SchoolContext();
            context.Subjects.Add(subject);
            context.SaveChanges();
        }

        public void Delete(string subjectId)
        {
            using var context = new SchoolContext();
            var subject = context.Subjects
                .First(s => s.SubjectId.Equals(subjectId));
            context.Remove(subject);
            context.SaveChanges();
        }

        public void Update(string subjectId, string name)
        {
            using var context = new SchoolContext();
            var subject = context.Subjects
                .First(s => s.SubjectId == subjectId);
            subject.Name = name;
            context.SaveChanges();
        }

        public IList<Subject> GetAll()
        {
            using var context = new SchoolContext();
            return context.Subjects.ToList();
        }
    }
}

File: Web Services and Middleware Technologies/Middleware/Pyro3Server/repository/subject_repository.py
from config.config import Session
from domain.Subject import Subject


class SubjectRepository:

    def __init__(self):
        self.__session = Session()

    def add(self, subject):
        try:
            self.__session.add(subject)
            self.__session.commit()
        except Exception:
            self.__session.rollback()
            raise Exception('Could not save subject!\n')

    def delete(self, subject_id):
        try:
            subject = self.__session.query(Subject).get(subject_id)
            self.__session.delete(subject)
            self.__session.commit()
        except Exception:
            self.__session.rollback()
            raise Exception('Could not delete subject!\n')

    def update(self, subject_id, name):
        try:
            field_map = {'name': name}
            self.__session.query(Subject) \
                .where(Subject.subject_id == subject_id) \
                .update(field_map)
            self.__session.commit()
        except Exception:
            self.__session.rollback()
            raise Exception('Could not update subject!\n')

    def get_all(self):
        return self.__session.query(Subject).all()
